trip_app/trips/views.py

```diff
@@ -69,35 +80,79 @@ def get_queryset(self):
 
     @action(detail=True, methods=["post"])
     def reserve(self, request, pk=None):
-        """ Reserve a trip by sending a POST to api/trip/<trip_id>/reserve/."""
+        """ Reserve a trip via POST api/trips/<trip_id>/reserve/. """
         trip = self.get_object()
         user = request.user
         if user == trip.driver:
             return Response(
                 data="Driver can't be a passenger at the same time",
                 status=status.HTTP_400_BAD_REQUEST,
             )
-        if trip.free_seats:
-            if trip.man_approve:
-                trip_request = TripRequest.objects.create(trip=trip, user=user)
-                serializer = TripRequestSerializer(trip_request)
-                return Response(
-                    serializer.data, status=status.HTTP_201_CREATED
-                )
-            trip.passengers.add(user)
-            serializer = self.get_serializer(trip)
-            return Response(serializer.data)
-        return Response(
-            data="There are no empty seats in this trip",
-            status=status.HTTP_400_BAD_REQUEST,
-        )
-
-    @action(
-        detail=True, methods=["get"], permission_classes=[IsTripDriverOrAdmin]
-    )
+        if user.requests.filter(trip_id=pk).exists():
+            return Response(
+                data="You have already requested this trip",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        trip_request = trip.process_request(user)
+        serializer = TripRequestSerializer(trip_request)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=["get"])
     def requests(self, *args, **kwargs):
         """ The list of the requests related to the specific trip.
             GET /api/trips/<trip_id>/requests/. """
         trip = self.get_object()
         serializer = TripRequestSerializer(trip.requests, many=True)
         return Response(serializer.data)
+
+
+class TripRequestViewSet(
+    mixins.RetrieveModelMixin, mixins.ListModelMixin, viewsets.GenericViewSet
+):
+    """ Provide listing or retrieving trip requests,
+        with the approve/decline/cancel trip request custom actions. """
+
+    serializer_class = TripRequestSerializer
+    queryset = TripRequest.objects.all()
+    pagination_class = PageNumberPagination
+    pagination_class.page_size = 5
+
+    def get_permissions(self):
+        if self.action in ["approve", "decline"]:
```

Comment:
Please get rid of multiple `if ... else` statements, try to use mapper to control cases.

Reply:
Thanks for good advice. This should be updated now.